doc/design/queue_api.md

```diff
@@ -0,0 +1,382 @@
+
+### Scope
+
+  * In MVP, the resource request of `QueueJob` are ignored
+  * In MVP, the policy allocates resource based on `Queue`'s configuration
+  * In MVP, `Queue` is namespace level
+
+## Terminology
+
+  * Deserved (Resource): The total number of resources that the arbitrator allocated to the namespace
+  * Overused: The namespace is overused if it used more resources than deserved resources
+  * Underused: The namespace is underused if it used less resource than deserved resources
+
+## API
+
+```go
+type Queue struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    Spec QueueSpec
+    Status QueueStatus
+}
+
+type QueueSpec struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    // The weight of Queue, which is used by policy to allocatre resource; the 
+    // default value is 1. NOTE: it can not expect allocating more resouce with 
+    // higher weithg, it dependent on policy's reaction to the weight.
+    Weight int
+}
+
+type QueueStatus struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    // The deserved resource of Queue according to the policy
+    Deserved    ResourceList
+    // The resources that allocated to Queue, the allocated resource is less or 
+    // equal to `Deserved`:
+    //   * if some resource was preempting, the Allocated is less then Deserved
+    //   * otherwise, Allocated equals to Deserved
+    Allocated   ResourceList
+    // The resource that used by Pod in namespace; if more resource was used than 
+    // Deserved, the overused resource will be preempted.
+    Used        ResourceList
+    // The resources that are preempting for Queue
+    Preempting  ResourceList
+}
+```
+
+## Function Detail
+
+### Workflow
+![workflow](../images/workflow.jpg)
+
+### Admission Controller
+
+### Quota Manager
+Only Quota Manager can update Resource Quota. And it has two responsibility:
+
+* Periodically query Queue status which contains allocated resources information from API server. Status struct is defined as `QueueStatus`
+* Update `Allocated ` information in `QueueStatus` into Resource Quota
+
+### Scheduler Cache
+
+Scheduler Cache periodically fetches all Node/Pod/Queue information in the cluster from API server. That inforamtion will only be stored in memory and not persisted on disk.
+
+It provides two interfaces `Run()` and `Dump()`
+
+* `Run()` to trigger cache to periodically fetch Node/Pod/Queue information from API server
+* `Dump()` create `CacheSnapshot` for policy
+
+```go
+type Cache interface {
+	// trigger cache to fetch Node/Pod/Queue 
+	// information periodically from API server
+	Run(stopCh <-chan struct{})
+
+	// Dump deep copy overall Node/Pod/Queue information into CacheSnapshot
+	Dump() *CacheSnapshot
+}
+
+type CacheSnapshot struct {
+	Pods       []*PodInfo
+	Nodes      []*NodeInfo
+	Queues     []*QueueInfo
+}
+```
+
```

Comment:
If schedulercache could be shared by queuejob controller, or potential other additional controllers, various "snapshot" may be needed. For example,  a "QueueJobCacheSnapshot" would contains PodInfo/QueueInfo/QueuejobInfo.

Reply:
I think we may add `QueuejobInfo` into `CacheSnapshot`. Controller dumps all snapshots and uses related information.
Maybe we can separate the snapshot later if dumping extra information is too heavy.